### PR TITLE
Get ready for v0.11 and v1.0

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
This PR catchup [Get ready for v0.11 and v1.0!](https://dev.embulk.org/topics/get-ready-for-v0.11-and-v1.0.html)

v0.10 API ready and also, still supports v0.9.

@dmikurube Could you take a look when you get a chance?

### CODE REVIEW

**Comment by hiroyuki-sato** [comment]
I have two issues.

## 1. CharsetModule is already added

```
2021-06-11 22:41:06.333 +0900 [WARN] (0001:transaction): Jackson Module class org.embulk.util.config.modules.CharsetModule is already added. It may be duplicated.
2021-06-11 22:41:06.392 +0900 [INFO] (0001:transaction): {done:  0 / 1, running: 0}
2021-06-11 22:41:06.417 +0900 [WARN] (0012:task-0000): Jackson Module class org.embulk.util.config.modules.CharsetModule is already added. It may be duplicated.
```

## 2. java.lang.ExceptionInInitializerError

```
org.embulk.exec.PartialExecutionException: java.lang.RuntimeException: java.lang.ExceptionInInitializerError
	at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(BulkLoader.java:344)
	at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:584)
	at org.embulk.exec.BulkLoader.access$000(BulkLoader.java:37)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:357)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:354)
	at org.embulk.spi.ExecInternal.doWith(ExecInternal.java:26)
	at org.embulk.exec.BulkLoader.run(BulkLoader.java:354)
	at org.embulk.EmbulkEmbed.run(EmbulkEmbed.java:324)
	at org.embulk.EmbulkRunner.runInternal(EmbulkRunner.java:288)
	at org.embulk.EmbulkRunner.run(EmbulkRunner.java:153)
	at org.embulk.cli.EmbulkRun.runInternal(EmbulkRun.java:119)
	at org.embulk.cli.EmbulkRun.run(EmbulkRun.java:24)
	at org.embulk.cli.Main.main(Main.java:55)
Caused by: java.lang.RuntimeException: java.lang.ExceptionInInitializerError
	at org.embulk.exec.BulkLoader$LoaderState.getRepresentativeException(BulkLoader.java:289)
	at org.embulk.exec.BulkLoader.execute(BulkLoader.java:702)
	at org.embulk.exec.BulkLoader.access$300(BulkLoader.java:37)
	at org.embulk.exec.BulkLoader$4$1$1$1.run(BulkLoader.java:546)
	at org.embulk.spi.FileOutputRunner$RunnerControl$1$1.run(FileOutputRunner.java:95)
	at org.embulk.formatter.csv.CsvFormatterPlugin.transaction(CsvFormatterPlugin.java:138)
	at org.embulk.spi.FileOutputRunner$RunnerControl$1.run(FileOutputRunner.java:90)
	at org.embulk.spi.util.EncodersInternal$RecursiveControl.transaction(EncodersInternal.java:68)
	at org.embulk.spi.util.EncodersInternal.transaction(EncodersInternal.java:29)
	at org.embulk.spi.FileOutputRunner$RunnerControl.run(FileOutputRunner.java:88)
	at org.embulk.output.S3FileOutputPlugin.resume(S3FileOutputPlugin.java:321)
	at org.embulk.output.S3FileOutputPlugin.transaction(S3FileOutputPlugin.java:314)
	at org.embulk.spi.FileOutputRunner.transaction(FileOutputRunner.java:59)
	at org.embulk.exec.BulkLoader$4$1$1.transaction(BulkLoader.java:539)
	at org.embulk.exec.LocalExecutorPlugin.transaction(LocalExecutorPlugin.java:51)
	at org.embulk.exec.BulkLoader$4$1.run(BulkLoader.java:534)
	at org.embulk.spi.util.FiltersInternal$RecursiveControl.transaction(FiltersInternal.java:84)
	at org.embulk.spi.util.FiltersInternal.transaction(FiltersInternal.java:42)
	at org.embulk.exec.BulkLoader$4.run(BulkLoader.java:529)
	at org.embulk.spi.FileInputRunner$RunnerControl$1$1.run(FileInputRunner.java:111)
	at org.embulk.parser.csv.CsvParserPlugin.transaction(CsvParserPlugin.java:309)
	at org.embulk.spi.FileInputRunner$RunnerControl$1.run(FileInputRunner.java:106)
	at org.embulk.spi.util.DecodersInternal$RecursiveControl.transaction(DecodersInternal.java:66)
	at org.embulk.spi.util.DecodersInternal$RecursiveControl$1.run(DecodersInternal.java:62)
	at org.embulk.decoder.gzip.GzipFileDecoderPlugin.transaction(GzipFileDecoderPlugin.java:38)
	at org.embulk.spi.util.DecodersInternal$RecursiveControl.transaction(DecodersInternal.java:58)
	at org.embulk.spi.util.DecodersInternal.transaction(DecodersInternal.java:28)
	at org.embulk.spi.FileInputRunner$RunnerControl.run(FileInputRunner.java:104)
	at org.embulk.input.file.LocalFileInputPlugin.resume(LocalFileInputPlugin.java:92)
	at org.embulk.input.file.LocalFileInputPlugin.transaction(LocalFileInputPlugin.java:85)
	at org.embulk.spi.FileInputRunner.transaction(FileInputRunner.java:61)
	at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:525)
	... 11 more
Caused by: java.lang.ExceptionInInitializerError
	at com.amazonaws.util.Md5Utils.md5AsBase64(Md5Utils.java:104)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1647)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.putFile(S3FileOutputPlugin.java:202)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.closeCurrent(S3FileOutputPlugin.java:212)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.finish(S3FileOutputPlugin.java:269)
	at org.embulk.util.file.FileOutputOutputStream.close(FileOutputOutputStream.java:111)
	at sun.nio.cs.StreamEncoder.implClose(StreamEncoder.java:320)
	at sun.nio.cs.StreamEncoder.close(StreamEncoder.java:149)
	at java.io.OutputStreamWriter.close(OutputStreamWriter.java:233)
	at java.io.BufferedWriter.close(BufferedWriter.java:266)
	at org.embulk.util.text.LineEncoder.finish(LineEncoder.java:116)
	at org.embulk.formatter.csv.CsvFormatterPlugin$1.finish(CsvFormatterPlugin.java:255)
	at org.embulk.spi.FileOutputRunner$DelegateTransactionalPageOutput.finish(FileOutputRunner.java:152)
	at org.embulk.spi.PageBuilderImpl.finish(PageBuilderImpl.java:241)
	at org.embulk.parser.csv.CsvParserPlugin.run(CsvParserPlugin.java:463)
	at org.embulk.spi.FileInputRunner.run(FileInputRunner.java:138)
	at org.embulk.spi.util.ExecutorsInternal.process(ExecutorsInternal.java:60)
	at org.embulk.spi.util.ExecutorsInternal.process(ExecutorsInternal.java:37)
	at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:171)
	at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:168)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at javax.xml.bind.ContextFinder.handleClassCastException(ContextFinder.java:129)
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:265)
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:249)
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:477)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:656)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:599)
	at com.amazonaws.util.Base64.<clinit>(Base64.java:44)
	... 24 more

Error: java.lang.RuntimeException: java.lang.ExceptionInInitializerError
```

**Comment by dmikurube** [comment]
Although I'm not yet sure of 1, I confirmed that 2 was resolved by updating by `com.amazonaws:aws-java-sdk-s3`.

----

Then, my suggestion is:
1. Release one version only with Embulk-level catch-up, without JAXB and without AWS SDK updates. It won't work with Java 9, but it should work with Embulk v0.10 and v0.11.
2. Next, release another version in addition with AWS SDK updates (to the latest `1.11.1034`), and JAXB. It should work with Java 9+, Embulk v0.10 and v0.11.

Relasing 1 first is to mitigate any possible compatibility issue.

----

Before releasing, we still need to struggle `1. CharsetModule is already added", though.

**Comment by dmikurube** [comment]
Ah, got it. I guess `1. CharsetModule is already added` does not look coming from this `embulk-output-s3`. It is from the CSV formatter plugin you use.

The standard plugins still use `embulk-util-config:0.2.1` until v0.10.31.
https://github.com/embulk/embulk/blob/v0.10.31/embulk-formatter-csv/src/main/java/org/embulk/formatter/csv/CsvFormatterPlugin.java#L373-L376

We plan to update `embulk-util-config` to `0.3.0` with removing unnecessary `Module`s along with Embulk v0.10.32.
https://github.com/embulk/embulk/pull/1457

**Comment by hiroyuki-sato** [comment]
@dmikurube Thank you for investigating the issues.

I would like to double-check your comment. Are these steps correct?

#### 1st version.

Remove explicit JAXB dependency. It is mean that remove the following parts from the current PR.
It will output the warning `Class javax.xml.bind.JAXBContext is loaded by the parent ClassLoader, which is removed by JEP 320`. But we can ignore it.

```java
// JAXB 2.2.11 is chosen here because:
// 1. JDK 8's bundled JAXB is 2.2.8. Better with a closer version while we are on Java 8.
//    https://javaee.github.io/jaxb-v2/doc/user-guide/ch02.html#a-2-2-8
// 2. Neither com.sun.xml.bind:jaxb-core:2.2.8 nor com.sun.xml.bind:jaxb-impl:2.2.8 does not exist on Maven Central.
// 3. 2.2.11 looks to be used by the most Java libraries among JAXB 2.2.
//    https://mvnrepository.com/artifact/javax.xml.bind/jaxb-api
//    https://mvnrepository.com/artifact/com.sun.xml.bind/jaxb-core
//    https://mvnrepository.com/artifact/com.sun.xml.bind/jaxb-impl
// 4. JAXB 2.2.8 and 2.2.11 look to have the same set of classes.
//    Although their internal implementations are a bit different, class loaders would not be confused.
compile "javax.xml.bind:jaxb-api:2.2.11"
compile "com.sun.xml.bind:jaxb-core:2.2.11"
compile "com.sun.xml.bind:jaxb-impl:2.2.11"
```

#### 2nd version

* Update AWS SDK to `1.11.1034` or newer `1.11.XXXX`.
* Re-enable JAXB dependency.

**Comment by dmikurube** [comment]
That's right!

**Comment by hiroyuki-sato** [comment]
@dmikurube I fixed your comment. This is the PR for 1st version.
It raises the following exception if a user uses this plugin with embulk v0.10.x

```
org.embulk.plugin.PluginClassLoader$DummyStackTraceDump: [DUMMY] It is not an Exception. Just showing where classes which will be removed in JEP 320 are loaded.
	at org.embulk.plugin.PluginClassLoader.logInfoIfJep320Class(PluginClassLoader.java:248)
	at org.embulk.plugin.PluginClassLoader.loadClass(PluginClassLoader.java:148)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at com.amazonaws.util.Base64.<clinit>(Base64.java:44)
	at com.amazonaws.util.Md5Utils.md5AsBase64(Md5Utils.java:104)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1647)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.putFile(S3FileOutputPlugin.java:203)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.closeCurrent(S3FileOutputPlugin.java:213)
	at org.embulk.output.S3FileOutputPlugin$S3FileOutput.finish(S3FileOutputPlugin.java:270)
	at org.embulk.util.file.FileOutputOutputStream.close(FileOutputOutputStream.java:111)
	at sun.nio.cs.StreamEncoder.implClose(StreamEncoder.java:320)
	at sun.nio.cs.StreamEncoder.close(StreamEncoder.java:149)
	at java.io.OutputStreamWriter.close(OutputStreamWriter.java:233)
	at java.io.BufferedWriter.close(BufferedWriter.java:266)
	at org.embulk.util.text.LineEncoder.finish(LineEncoder.java:116)
	at org.embulk.formatter.csv.CsvFormatterPlugin$1.finish(CsvFormatterPlugin.java:255)
	at org.embulk.spi.FileOutputRunner$DelegateTransactionalPageOutput.finish(FileOutputRunner.java:152)
	at org.embulk.exec.LocalExecutorPlugin$ScatterTransactionalPageOutput.finish(LocalExecutorPlugin.java:445)
	at org.embulk.spi.PageBuilderImpl.finish(PageBuilderImpl.java:241)
	at org.embulk.parser.csv.CsvParserPlugin.run(CsvParserPlugin.java:463)
	at org.embulk.spi.FileInputRunner.run(FileInputRunner.java:138)
	at org.embulk.exec.LocalExecutorPlugin$ScatterExecutor.runInputTask(LocalExecutorPlugin.java:270)
	at org.embulk.exec.LocalExecutorPlugin$ScatterExecutor.access$100(LocalExecutorPlugin.java:195)
	at org.embulk.exec.LocalExecutorPlugin$ScatterExecutor$1.call(LocalExecutorPlugin.java:234)
	at org.embulk.exec.LocalExecutorPlugin$ScatterExecutor$1.call(LocalExecutorPlugin.java:231)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by dmikurube** [comment]
As you can see the message:

```
[DUMMY] It is not an Exception. Just showing where classes which will be removed in JEP 320 are loaded.
```

That is not a real Exception.

**Comment by hiroyuki-sato** [comment]
@dmikurube Thanks. You are right. It's not an exception. just a warning.
I just want to record that this PR outputs this **warning** if a user uses this plugin with Embulk v0.10.x.

File: build.gradle [hunk 1]
```diff
@@ -1,36 +1,94 @@
...
-import com.github.jrubygradle.JRubyExec
 repositories {
     mavenCentral()
-    jcenter()
-}
-configurations {
-    provided
 }
 
+group = "io.github.llibra"
```
Comment (dmikurube) [inline build.gradle:12]: Let's update this to `org.embulk` now.

File: build.gradle [hunk 2]
```diff
@@ -1,36 +1,94 @@
 plugins {
-    id "com.jfrog.bintray" version "1.1"
-    id "com.github.jruby-gradle.base" version "0.1.5"
     id "java"
     id "checkstyle"
+    id "maven-publish"
+//    id "signing"
```
Comment (dmikurube) [inline build.gradle:5]: Let's have `"signing"`.

File: build.gradle [hunk 3]
```diff
@@ -1,36 +1,94 @@
...
+        exclude group: "com.fasterxml.jackson.core", module: "jackson-annotations"
+        exclude group: "com.fasterxml.jackson.core", module: "jackson-core"
+        exclude group: "com.fasterxml.jackson.core", module: "jackson-databind"
+        exclude group: "joda-time", module: "joda-time"
+
+        // commons-logging api is provided by jcl-over-slf4j below.
+        exclude group: "commons-logging", module: "commons-logging"
+    }
+
+    compile("org.embulk:embulk-util-config:0.3.0") {
```
Comment (dmikurube) [inline build.gradle:35]: Let's have `0.3.1` now -- no impacting changes, but we've improved its Exception messages.
https://github.com/embulk/embulk-util-config/pull/30

File: build.gradle [hunk 4]
```diff
@@ -1,36 +1,94 @@
...
-    doFirst { file("classpath").deleteDir() }
-    from (configurations.runtime - configurations.provided + files(jar.archivePath))
-    into "classpath"
+    // They are once excluded from transitive dependencies of other dependencies,
+    // and added explicitly with versions exactly the same with embulk-core:0.10.19.
+    compile "com.fasterxml.jackson.core:jackson-annotations:2.6.7"
+    compile "com.fasterxml.jackson.core:jackson-core:2.6.7"
+    // embulk-output-s3 1.5.0 contains jackson-databind 2.6.7.1
+    // in the gem file even embulk-core uses 2.6.7.
+    // So, I write here exactly the same version explicitly for keep compatibility.
```
Comment (dmikurube) [inline build.gradle:52]:
```suggestion
    // "jackson-databind:2.6.7.1" is a little bit different from "embulk-core:0.10.31"'s dependency ("2.6.7"). But,
    // "embulk-output-s3" had contained "jackson-databind:2.6.7.1", not "2.6.7", for a long time. It has been working.
    // Keeping this "2.6.7.1" here for compatibility with older versions. It has no problems with Embulk v0.10.32+.
```

File: build.gradle [hunk 5]
```diff
@@ -1,36 +1,94 @@
...
 repositories {
     mavenCentral()
-    jcenter()
-}
-configurations {
-    provided
 }
 
+group = "io.github.llibra"
 version = "1.5.0"
```
Comment (dmikurube) [inline build.gradle:13]: Maybe nice to upgrade to the next `1.6.0` ?

File: build.gradle [hunk 6]
```diff
@@ -1,36 +1,94 @@
...
+
+        // slf4j-api conflicts with embulk-api. It should be excluded.
+        exclude group: "org.slf4j", module: "slf4j-api"
+    }
+
+    // TODO: Uncomment compile statement when updates aws-java-sdk-s3.
+    //       Current aws-java-sdk-s3:1.11.27 raises the following exception
+    //       java.lang.ExceptionInInitializerError
+    //       if write JAXB dependency explicitly.
+    //
```
Comment (dmikurube) [inline build.gradle:70]:
```suggestion
    // TODO: Enable those JAXB dependencies.
    //
    // Including explicit JAXB dependencies would be needed to work with Java 9+. But they didn't work with aws-java-sdk-s3:1.11.271.
    // On the other hand, upgrading aws-java-sdk has another compatibility risk.
    //
    // We decided to release one version without JAXB and AWS SDK upgrade, but only with v0.10 catch-up, at first.
    // Then, we will release a next version both with JAXB and AWS SDK upgrade.
    //
```

File: build.gradle [hunk 7]
```diff
@@ -49,29 +107,61 @@ task checkstyle(type: Checkstyle) {
...
+    category = "output"
+    type = "s3"
+}
 
-  spec.add_development_dependency 'bundler', ['~> 1.0']
-  spec.add_development_dependency 'rake', ['>= 10.0']
-end
-/$)
+gem {
+    from("LICENSE.txt")  // Optional -- if you need other files in the gem.
```
Comment (dmikurube) [inline build.gradle:153]: Let's remove this for the same reason with the above.

File: build.gradle [hunk 8]
```diff
@@ -49,29 +107,61 @@ task checkstyle(type: Checkstyle) {
...
+    authors = [ "Manabu Takayama" ]
+    email = [ "learn.libra@gmail.com" ]
+    // "description" of the gem is copied from "description" of your Gradle project.
+    summary = "S3 file output plugin for Embulk"
+    homepage = "https://github.com/llibra/embulk-output-s3"
```
Comment (dmikurube) [inline build.gradle:158]: Let's update this URL.

File: build.gradle [hunk 9]
```diff
@@ -49,29 +107,61 @@ task checkstyle(type: Checkstyle) {
...
+    from rootProject.file("LICENSE")
+}
+
+sourcesJar {
+    from rootProject.file("LICENSE")
+}
+
+javadocJar {
+    from rootProject.file("LICENSE")
 }
```
Comment (dmikurube) [inline build.gradle:125]: Let's remove them at the moment.

This plugin has been published under the MIT License, and it had `LICENSE.txt`. But, in another PR, I think about switching this to the Apache License 2.0 as other `github.com/embulk` plugins do.

**Review by dmikurube — COMMENTED** [review]
Left some more comments to be alinged with the ownership transfer.

File: build.gradle (hunk 1 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Changed to `org.embulk`.

File: build.gradle (hunk 2 above)
Comment (hiroyuki-sato) [reply build.gradle]: Enable `signing`

There is no configuration about Maven Central yet.
So I don't enable the following parts yet.

```java
// TODO: Enable those settings after append maven configuration.
//signing {
//    sign publishing.publications.maven
//}
```

File: build.gradle (hunk 3 above)
Comment (hiroyuki-sato) [reply build.gradle]: Fixed

File: build.gradle (hunk 4 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Changed.

File: build.gradle (hunk 6 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Changed.

File: build.gradle (hunk 9 above)
Comment (hiroyuki-sato) [reply build.gradle]: Removed `from rootProject.file("LICENSE")` parts.

File: build.gradle (hunk 8 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Updated.

File: build.gradle (hunk 7 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Removed

File: build.gradle (hunk 5 above)
Comment (hiroyuki-sato) [reply build.gradle]: Thanks. Updated to `1.6.0`

**Comment by hiroyuki-sato** [comment]
@dmikurube Thank you for your review. I fixed your comments.
Please let me know if I need to add the Maven Central configuration.
I think It is better to add after change the License.

**Review by dmikurube — APPROVED** [review]
LGTM! Let me merge this. I think I'll make some follow-up PRs to release 1.6.0.

**Pull request merged**